### PR TITLE
DynamoDb support

### PR DESCRIPTION
**Pull request by mikemoody-amido**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/MTTL-413

### *What is the problem we're trying to solve*

We need to add support for using DynamoDb to the Base API project template.

### *What changes have we introduced*

* Support for setting up DynamoDb as a database, including all necessary DI registrations.
* An example Gateway implementation to show how to use the DynamoDb context interface.
* DynamoDb setup for integration tests and an example integration test file to show how it works.
* Sample Terraform files to setup a DynamoDb instance and create the necessary table(s)

#### _Checklist_

- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Update the relevant wiki pages in the repo for the new functionality.

### CODE REVIEW

File: BaseApi/V1/Infrastructure/DynamoDbInitilisationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
+using Amazon.DynamoDBv2;
+using Amazon.DynamoDBv2.DataModel;
+using Microsoft.Extensions.DependencyInjection;
+using System;
+
+namespace BaseApi.V1.Infrastructure
+{
+    public static class DynamoDbInitilisationExtensions
+    {
+        public static void ConfigureDynamoDB(this IServiceCollection services)
```
Comment (rocco-scaramuzzi) [inline BaseApi/V1/Infrastructure/DynamoDbInitilisationExtensions.cs:10]: Is this used only by the test project?

File: BaseApi.Tests/DynamoDbIntegrationTests.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    public class DynamoDbIntegrationTests<TStartup> where TStartup : class
+    {
+        protected HttpClient Client { get; private set; }
+        private DynamoDbMockWebApplicationFactory<TStartup> _factory;
+        protected IDynamoDBContext DynamoDbContext => _factory?.DynamoDbContext;
+        protected List<Action> CleanupActions { get; set; }
+
+        private readonly List<TableDef> _tables = new List<TableDef>
+        {
+            // TODO: Populate the list of table(s) and their key property details here, for example:
```
Comment (evak6979) [inline BaseApi.Tests/DynamoDbIntegrationTests.cs:19]: Is this a forgotten comment?

File: BaseApi.Tests/V1/E2ETests/ExampleDynamoDbTest.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+using BaseApi.V1.Factories;
+using BaseApi.V1.Infrastructure;
+using FluentAssertions;
+using Newtonsoft.Json;
+using NUnit.Framework;
+using System;
+using System.Net;
+using System.Threading.Tasks;
+
+//namespace BaseApi.Tests.V1.E2ETests
```
Comment (evak6979) [inline BaseApi.Tests/V1/E2ETests/ExampleDynamoDbTest.cs:14]: Is this file needed since it's all commented out.

File: BaseApi/terraform/staging/main.tf [hunk 4]
```diff
@@ -60,3 +60,38 @@ module "development" {
   task_definition_secrets      = {}
   task_definition_secret_count = number # This number needs to reflect the number of environment variables provided
 }
+
+# THE SECTION BELOW IS FOR USE WITH DYNAMODB
```
Comment (evak6979) [inline BaseApi/terraform/staging/main.tf:64]: All these sections commented - should we keep them or uncomment them? (sorry, new to this version of organizing DynamoDB)

File: BaseApi/V1/Infrastructure/DynamoDbInitilisationExtensions.cs (hunk 1 above)
Comment (mikemoody-amido) [reply BaseApi/V1/Infrastructure/DynamoDbInitilisationExtensions.cs:10]: No. It's also in the api project `Startup.ConfigureServices()` method, but commented out and with a comment to use it if DynamoDb is being used.

File: BaseApi.Tests/DynamoDbIntegrationTests.cs (hunk 2 above)
Comment (mikemoody-amido) [reply BaseApi.Tests/DynamoDbIntegrationTests.cs:19]: No. When running integration tests the service expects the tables to be pre-created, so within the test project startup this list is used to ensure that any required tables exist before any tests are run.

File: BaseApi.Tests/V1/E2ETests/ExampleDynamoDbTest.cs (hunk 3 above)
Comment (mikemoody-amido) [reply BaseApi.Tests/V1/E2ETests/ExampleDynamoDbTest.cs:14]: It's is there as an example of how to write integration tests if using DynamoDb. If Postgres is being used then the dev can delete it. It's all commented out because if not then the tests will not work out of the box without changing it the whole thing from using Postgres (which is still the default) to DynamoDb.

File: BaseApi/terraform/staging/main.tf (hunk 4 above)
Comment (mikemoody-amido) [reply BaseApi/terraform/staging/main.tf:64]: New to me too.
I asked @LBHMGeorgieva about that. She advised me that it would be better to have a single template project that could support both Postgres and DynamoDb, but that the DynamoDb stuff would initially be commented out. A dev creating a new Api from the template could then use one and delete the other as needed.

File: BaseApi/terraform/production/main.tf [hunk 5]
```diff
@@ -60,3 +60,38 @@ module "development" {
   task_definition_secrets      = {}
   task_definition_secret_count = number # This number needs to reflect the number of environment variables provided
 }
+
```
Comment (LBHMGeorgieva) [inline BaseApi/terraform/production/main.tf:63]: This file will stay the same as it is, perhaps we can just delete the 'development' module from it as this is just the file specifying the provider and setting up the s3 state (so the commented out puts are not required). This applies to the other main.tf files as well.

**Review by LBHMGeorgieva — APPROVED** [review]
Please remove all DynamoDB terraform files. I will add a small task to add those to a template so over here we can just be creating a small module for each DynamoDB table provisioned. When the template is ready, we can update the base API to be making use of it.

File: BaseApi/terraform/production/main.tf (hunk 5 above)
Comment (mikemoody-amido) [reply BaseApi/terraform/production/main.tf:63]: I will remove all change I made to the main.tf files to leave them as they were.
